src/submission/validation.ts

```diff
@@ -140,6 +146,54 @@ export const usingInvalidProgramId = (
   }
   return [];
 };
+
+export const checkUniqueRecords = (
+  clinicalType: ClinicalEntityType,
+  newRecords: DeepReadonly<DataRecord[] | CreateRegistrationRecord[]>,
+  useAllRecordValues?: boolean, // use all record properties so it behaves like duplicate check
+): SubmissionValidationError[] => {
+  const identifierToIndexMap: { [k: string]: number[] } = {};
+  const indexToErrorMap: { [index: number]: SubmissionValidationError } = {};
+  const uniqueIdName = ClinicalUniqueIndentifier[clinicalType];
+
+  newRecords.forEach((record: any, index) => {
+    const uniqueIdentiferValue = useAllRecordValues ? JSON.stringify(record) : record[uniqueIdName];
+
+    if (!identifierToIndexMap[uniqueIdentiferValue]) {
```

Comment:
a general comment instead of if/else use if / return and decompose big functions into smaller private functions to be easier to follow

Reply:
:+1: